File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
 bool DFS(vector<int>adj[],vector<int>&vis,vector<int>&pathvis,int node){
    vis[node] = 1;
    pathvis[node] = 1;
    for(auto it : adj[node]){
       if(vis[it] == 1 && pathvis[it] == 1) return true;
       else if(!vis[it]){
        if(DFS(adj,vis,pathvis,it)) return true;
       }
    }
    pathvis[node] = 0;

    return false;
 }
    vector<int> findOrder(int V, vector<vector<int>>& pre) {
         vector<int>adj[V];
        vector<int>indeg(V,0);
        vector<int>vis(V,0);
        vector<int>pathvis(V,0);
        vector<int>ans;
        for(int i=0;i<pre.size();i++){
          adj[pre[i][1]].push_back(pre[i][0]);
          
        }

        for(int i=0;i<V;i++){
            if(!vis[i]){
            if(DFS(adj,vis,pathvis,i)) return ans;
          }
        }

        for(int i=0;i<V;i++){
            for(auto it : adj[i]){
                indeg[it]++;
            }
        }
        queue<int>q;
        for(int i=0;i<indeg.size();i++){
          if(indeg[i] == 0){
            q.push(i);
          }
        }
        while(!q.empty()){
            int node = q.front();
            q.pop();
            ans.push_back(node);

            for(auto it : adj[node]){
               indeg[it]--;
               if(indeg[it] == 0) q.push(it);
            }
        }
        return ans;
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
     boolean DFS_loop(List<List<Integer>> ls,int src,int[] vis, int[] pathvis){
      vis[src] = 1;
      pathvis[src] = 1;

      for(int i=0;i<ls.get(src).size();i++){
          int node = ls.get(src).get(i);
          if(vis[node] == 0){
            if(DFS_loop(ls,node,vis,pathvis)) return true;
          }
          else if(pathvis[node] == 1) return true;
      }
      pathvis[src] = 0;
      return false;
    }
    void DFS(List<List<Integer>> ls,int src,int[] vis, Stack<Integer>st){
    vis[src] = 1;

     for(int i=0;i<ls.get(src).size();i++){
          int node = ls.get(src).get(i);
          if(vis[node] == 0){
            DFS(ls,node,vis,st);
          }
          
      }
      st.push(src);
    }
    public int[] findOrder(int num, int[][] grid) {
        List<List<Integer>> ls = new ArrayList<>();
        int[] vis = new int[num];
        int[] pathvis = new int[num];
        Stack<Integer> st = new Stack<>();
        Arrays.fill(vis,0);
         Arrays.fill(pathvis,0);
        
        for(int i=0;i<num;i++){
          ls.add(new ArrayList<>());
        }
        for(int i=0;i<grid.length;i++){
          ls.get(grid[i][1]).add(grid[i][0]);
        }
        
        for(int i=0;i<num;i++){
          if(vis[i] == 0){
            if(DFS_loop(ls,i,vis,pathvis)) return new int[0];
          }
        }
        Arrays.fill(vis,0);
        for(int i=0;i<num;i++){
          if(vis[i] == 0){
            DFS(ls,i,vis,st);
          }
        }

        int[] ans = new int[st.size()];
        int idx = 0;
        while(st.size()>0){
            ans[idx++] = st.pop();
        }
        return ans;
    }
}